### PR TITLE
new resource: aws_ivs_channel

### PR DESCRIPTION
**Pull request by KevinEady**
### Description

Add new resource for AWS IVS Channel: aws_ivs_channel. This PR is the third in a set to support the full AWS IVS entities:

1. ~~Playback Key Pair~~
2. ~~Recording Configuration~~
3. Channel
4. Stream Key

### Relations

Relates #17272

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccIVSChannel_ PKG=ivs      
==> Checking that code complies with gofmt requirements...
TF_ACC=1 /Users/eakevinh/go/bin/go1.18.4 test ./internal/service/ivs/... -v -count 1 -parallel 20 -run='TestAccIVSChannel_'  -timeout 180m
=== RUN   TestAccIVSChannel_basic
=== PAUSE TestAccIVSChannel_basic
=== RUN   TestAccIVSChannel_tags
=== PAUSE TestAccIVSChannel_tags
=== RUN   TestAccIVSChannel_update
=== PAUSE TestAccIVSChannel_update
=== RUN   TestAccIVSChannel_disappears
=== PAUSE TestAccIVSChannel_disappears
=== RUN   TestAccIVSChannel_recordingConfiguration
=== PAUSE TestAccIVSChannel_recordingConfiguration
=== CONT  TestAccIVSChannel_basic
=== CONT  TestAccIVSChannel_disappears
=== CONT  TestAccIVSChannel_update
=== CONT  TestAccIVSChannel_recordingConfiguration
=== CONT  TestAccIVSChannel_tags
--- PASS: TestAccIVSChannel_disappears (24.90s)
--- PASS: TestAccIVSChannel_basic (34.09s)
--- PASS: TestAccIVSChannel_recordingConfiguration (48.65s)
--- PASS: TestAccIVSChannel_update (54.51s)
--- PASS: TestAccIVSChannel_tags (70.80s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ivs        73.610s
```

### CODE REVIEW

File: internal/service/ivs/channel.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+	if len(tags) > 0 {
+		in.Tags = Tags(tags.IgnoreAWS())
+	}
+
+	if v, ok := d.GetOk("type"); ok {
+		in.Type = aws.String(v.(string))
+	}
+
+	out, err := conn.CreateChannelWithContext(ctx, in)
```
Comment (KevinEady) [inline internal/service/ivs/channel.go:127]: The [`CreateChannelOutput`](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.44.133/service/ivs#CreateChannelOutput) contains both a `Channel` and `StreamKey` fields, however the [`GetChannelOutput`](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.44.133/service/ivs#GetChannelOutput) only contains the `Channel`. This has repercussions in the "return read after create" paradigm. I am unsure if the `aws_ivs_channel` schema should also have a `stream_key` computed field which could be populated by calling `ListStreamKeys` and `GetStreamKey`. 

The fourth and final PR of the set will introduce an `aws_ivs_stream_key` data source which retrieves this information via the above API calls, so a Terraform file could look something like:

```terraform
resource "aws_ivs_channel" "channel" {
  name = "channel-1"
}

data "aws_ivs_stream_key" "stream_key" {
  channel_arn = aws_ivs_channel.channel.arn
}
```

Just not sure if the key information should also be part of the channel, considering odds are that a user would generally want the stream key information on channel creation as well.

What are your thoughts?

**Review by KevinEady — COMMENTED** [review]
I have some implementation questions that I wanted to ask which may require some rework, so I wanted to get this PR tonight to grab some feedback. Please take a look. Thanks!

File: internal/service/ivs/channel.go (hunk 1 above)
Comment (ewbankkit) [reply internal/service/ivs/channel.go:127]: If `StreamKey` isn't used elsewhere then I would suggest leaving it out.
Having attributes set just in `Create` prevents resource import working correctly.

File: internal/service/ivs/status.go [hunk 2]
```diff
@@ -42,3 +44,29 @@ func statusRecordingConfiguration(ctx context.Context, conn *ivs.IVS, id string)
...
+			return out, statusNormal, nil
+		} else {
+			if (updateDetails.Authorized != nil && aws.BoolValue(updateDetails.Authorized) == aws.BoolValue(out.Authorized)) ||
+				(updateDetails.LatencyMode != nil && aws.StringValue(updateDetails.LatencyMode) == aws.StringValue(out.LatencyMode)) ||
+				(updateDetails.Name != nil && aws.StringValue(updateDetails.Name) == aws.StringValue(out.Name)) ||
+				(updateDetails.RecordingConfigurationArn != nil && aws.StringValue(updateDetails.RecordingConfigurationArn) == aws.StringValue(out.RecordingConfigurationArn)) ||
+				(updateDetails.Type != nil && aws.StringValue(updateDetails.Type) == aws.StringValue(out.Type)) {
+				return out, statusUpdated, nil
+			}
+			return out, statusChangePending, nil
```
[the remark this answers is not in the document]
Comment (ewbankkit) [reply internal/service/ivs/status.go:69]: Effectively, yes.
We do something similar for SQS queue attributes: https://github.com/hashicorp/terraform-provider-aws/blob/01599517f39e2db44b22b2e2f3896814c69387e9/internal/service/sqs/status.go#L29.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccIVSChannel_' PKG=ivs ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ivs/... -v -count 1 -parallel 2  -run=TestAccIVSChannel_ -timeout 180m
=== RUN   TestAccIVSChannel_basic
=== PAUSE TestAccIVSChannel_basic
=== RUN   TestAccIVSChannel_tags
=== PAUSE TestAccIVSChannel_tags
=== RUN   TestAccIVSChannel_update
=== PAUSE TestAccIVSChannel_update
=== RUN   TestAccIVSChannel_disappears
=== PAUSE TestAccIVSChannel_disappears
=== RUN   TestAccIVSChannel_recordingConfiguration
=== PAUSE TestAccIVSChannel_recordingConfiguration
=== CONT  TestAccIVSChannel_basic
=== CONT  TestAccIVSChannel_disappears
--- PASS: TestAccIVSChannel_disappears (18.36s)
=== CONT  TestAccIVSChannel_update
--- PASS: TestAccIVSChannel_basic (23.41s)
=== CONT  TestAccIVSChannel_tags
--- PASS: TestAccIVSChannel_update (34.64s)
=== CONT  TestAccIVSChannel_recordingConfiguration
--- PASS: TestAccIVSChannel_tags (46.56s)
--- PASS: TestAccIVSChannel_recordingConfiguration (29.26s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ivs	86.816s
```

**Comment by ewbankkit** [comment]
@KevinEady Thanks for the contribution :tada: :clap:.
If you are happy with the comments I can go ahead and merge this.

**Comment by KevinEady** [comment]
Hi @ewbankkit , 

Thanks for the review! Comments sound good to me, so this can go ahead and be merged.

I'll be pushing the final entity, `aws_ivs_stream_key` data source, on Monday. Enjoy your weekend!

Thanks, Kevin

**Pull request merged**